README
NAME
    Cocoa::Growl - Yet another growl module using Growl.framework

SYNOPSIS
        use Cocoa::Growl ':all';
    
        my $installed = growl_installed(); # true if Growl is installed.
        my $running   = growl_running();   # true if Growl is running.
    
        # register application
        growl_register(
            app           => 'My growl script',
            icon          => '/path/to/icon.png', # or 'http://url/to/icon'
            notifications => [qw(Notification1 Notification2)],
        );
    
        # show growl notification
        growl_notify(
            name        => 'Notification1',
            title       => 'Hello!',
            description => 'Growl world!',
        );

DESCRIPTION
FUNCTIONS
    No function is exported by default, but all functions is exportable. And
    ':all' tag export all functions.

  growl_installed
        my $installed = growl_installed();

    Return true value if growl is installed.

  growl_running
        my $running = growl_running();

    Return true value if growl is running.

  growl_register(%parameters)
    Register application to growl.

        growl_register(
            app           => 'My growl script',
            icon          => '/path/to/icon.png', # or 'http://url/to/icon'
            notifications => [qw(Notification1 Notification2)],
        );

    Available parameters are:

    *   app => 'Str' (Required)

        The name of the application. This is listed in Growl preference
        panel.

    *   icon => 'Str'

        Application icon image path or URL. This image is showed in Growl
        preference panel, and used notification default image.

    *   notifications => 'ArrayRef' (Required)

        List of notification names. These names will be displayed in Growl
        preference pane to let users customize options for each
        notification.

    *   defaults => 'ArrayRef'

        List of notification names to enable by default. If this parameter
        is not set, all notifications is to become default.

  growl_notify(%parameters)
    Show growl notify.

        growl_notify(
            name        => 'Notification1',
            title       => 'Hello!',
            description => 'Growl world!',
        );

    Available options are:

    *   name => 'Str' (Required)

        The internal name of the notification. Should be human-readable, as
        it will be displayed in the Growl preference pane. And this value is
        required to be registered by "growl_register" before calling this
        function.

    *   title => 'Str'

        The title of the notification displayed to the user.

    *   description => 'Str'

        The full description of the notification displayed to the user.

    *   icon => 'Str'

        Image file path or URL to show with the notification as its icon. If
        this value is not set, the application's icon will be used instead.

    *   sticky => 'Bool'

        If true value is set, the notification will remain on screen until
        clicked. Not all Growl displays support sticky notifications.

    *   priority => 'Int'

        The priority of the notification. The default value is 0; positive
        values are higher priority and negative values are lower priority.
        Not all Growl displays support priority.

    *   on_click => 'CodeRef',

        This callback is called when notification is clicked. See also
        CALLBACK NOTICE below.

    *   on_timeout => 'CodeRef',

        This callback is called when notification is timeout. (also called
        notification closed by close button)

   CALLBACK NOTICE
    You should run Cocoa's event loop NSRunLoop to be enable callbacks.
    Simplest way to do that is use this module with Cocoa::EventLoop.

        use Cocoa::EventLoop;
        use Cocoa::Growl ':all';
    
        growl_register(
            name          => 'test script',
            notifications => ['test notification'],
        );
    
        my $wait = 1;
        growl_notify(
            name        => 'test notification',
            title       => 'Hello',
            description => 'Growl World!',
            onClick => sub {
                warn 'click';
                $wait = 0;
            },
            onTimeout => sub {
                warn 'timeout';
                $want = 0;
            },
        );
    
        Cocoa::EventLoop->run_while(0.1) while unless $wait;

    If you want to write more complicated script, use AnyEvent and
    AnyEvent::Impl::NSRunLoop. AnyEvent::Impl::NSRunLoop is a wrapper for
    AnyEvent and Cocoa::EventLoop and by using this module, you can use
    cocoa's event loop transparently in your AnyEvent application.

    This is little example:

        use AnyEvent;
        use AnyEvent::Impl::NSRunLoop;
    
        use Cocoa::Growl ':all';;
    
        my $cv = AnyEvent->condvar;
    
        growl_register(
            name          => 'test script',
            notifications => ['test notification'],
        );
    
        growl_notify(
            name        => 'test notification',
            title       => 'Hello',
            description => 'Growl World!',
            onClick => sub {
                warn 'click';
                $cv->send;
            },
            onTimeout => sub {
                warn 'timeout';
                $cv->send;
            },
        );
    
        $cv->recv;

    This script show one notification and wait until notification closed.

USE YOUR OWN Growl.framework
    Although this module bundle Growl.framework and load it by default, you
    can load your own Growl.framework. To do that, save your Growl.framework
    to "/Library/Frameworks/Growl.framework/", and add
    "USE_LOCAL_GROWL_FRAMEWORK=1" option when run Makefile.PL

        perl Makefile.PL USE_LOCAL_GROWL_FRAMEWORK=1

AUTHOR
    Daisuke Murase <typester@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (c) 2010 by KAYAC Inc.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl extension Cocoa::Growl

0.04  2010-12-27T10:24:54+09:00
      - added missing dependency, File::ShareDir.

0.03  2010-12-17T13:38:18+09:00
      - fix compilation error with OSX system perl.

0.02  2010-12-16T20:58:09+09:00
      - fix documents

0.01  2010-12-16T18:10:52+09:00
      - initial release



growl.m
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"
#include "ppport.h"

#import <Foundation/Foundation.h>
#ifdef USE_LOCAL_GROWL
#import <Growl/Growl.h>
#else
#import <Growl.h>
#endif
#import <objc/runtime.h>

static Class GrowlAppBridge;

@interface GrowlContext : NSObject {
@public
    SV* timeout_cb;
    SV* click_cb;
}
@end

@implementation GrowlContext
-(void)dealloc {
    if (NULL != timeout_cb) SvREFCNT_dec(timeout_cb);
    if (NULL != click_cb) SvREFCNT_dec(click_cb);
    [super dealloc];
}
@end

@interface Growl : NSObject <GrowlApplicationBridgeDelegate> {
    NSDictionary* info_;
    NSMutableDictionary* contexts_;
}
+(Growl*)sharedInstance;
-(NSDictionary*)info;
-(void)setInfo:(NSDictionary*)newInfo;
-(NSString*)addContext:(GrowlContext*)context;
-(void)removeContext:(NSString*)key;
@end

@implementation Growl

+(Growl*)sharedInstance {
    static Growl* obj = nil;
    if (nil == obj) {
        obj = [[Growl alloc] init];
    }
    return obj;
}

-(id)init {
    if (self = [super init]) {
        info_ = nil;
        contexts_ = [[NSMutableDictionary dictionary] retain];
    }
    return self;
}

-(NSDictionary*)info {
    return info_;
}

-(void)setInfo:(NSDictionary*)newInfo {
    if (nil != info_) {
        [info_ release];
    }
    info_ = [newInfo retain];
}

-(NSString*)addContext:(GrowlContext*)context {
    // create dic key
    CFUUIDRef uuid_ref = CFUUIDCreate(nil);
    NSString* uuid = (NSString*)CFUUIDCreateString(nil, uuid_ref);
    CFRelease(uuid_ref);

    [contexts_ setObject:context forKey:uuid];

    return [uuid autorelease];
}

-(void)removeContext:(NSString*)key {
    [contexts_ removeObjectForKey:key];
}

-(NSDictionary*)registrationDictionaryForGrowl {
    return info_;
}

-(void)growlNotificationWasClicked:(id)clickContext {
    GrowlContext* context = [contexts_ objectForKey:(NSString*)clickContext];
    if (context && context->click_cb) {
        dSP;

        ENTER;
        SAVETMPS;

        PUSHMARK(SP);
        PUTBACK;

        call_sv(context->click_cb, G_SCALAR);

        SPAGAIN;

        PUTBACK;
        FREETMPS;
        LEAVE;
    }

    [self removeContext:clickContext];
    [context release];
}

-(void)growlNotificationTimedOut:(id)clickContext {
    GrowlContext* context = [contexts_ objectForKey:(NSString*)clickContext];
    if (context && context->timeout_cb) {
        dSP;

        ENTER;
        SAVETMPS;

        PUSHMARK(SP);
        PUTBACK;

        call_sv(context->timeout_cb, G_SCALAR);

        SPAGAIN;

        PUTBACK;
        FREETMPS;
        LEAVE;
    }

    [self removeContext:clickContext];
    [context release];
}

-(void)dealloc {
    [info_ release];
    [contexts_ release];
    [super dealloc];
}

@end

XS(growl_installed) {
    dXSARGS;

    SV* sv;

    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
    if ([GrowlAppBridge isGrowlInstalled]) {
        sv = newSViv(1);
    }
    else {
        sv = newSViv(0);
    }
    [pool drain];

    ST(0) = sv_2mortal(sv);

    XSRETURN(1);
}

XS(growl_running) {
    dXSARGS;

    SV* sv;

    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
    if ([GrowlAppBridge isGrowlRunning]) {
        sv = newSViv(1);
    }
    else {
        sv = newSViv(0);
    }
    [pool drain];

    ST(0) = sv_2mortal(sv);

    XSRETURN(1);
}

XS(growl_register) {
    dXSARGS;

    if (items < 3) {
        Perl_croak(aTHX_ "Usage: _growl_register($app_name, \\@allNotifications, \\@defaultNotifications[, $iconFile])");
    }

    SV* sv_appName = ST(0);
    SV* sv_allNotif = ST(1);
    SV* sv_defaultNotif = ST(2);
    SV* sv_image = NULL;
    if (items >= 4) sv_image = ST(3);

    if (!SvROK(sv_allNotif) || !SvROK(sv_defaultNotif) ||
        SVt_PVAV != SvTYPE(SvRV(sv_allNotif)) ||
        SVt_PVAV != SvTYPE(SvRV(sv_defaultNotif))) {

        Perl_croak(aTHX_ "Error: \\@allNotifications and \\@defaultNotifications should be ArrayRef");
    }

    STRLEN len;
    char* ptr;

    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];

    ptr = SvPV(sv_appName, len);
    NSString* appName = [NSString stringWithUTF8String:ptr];

    NSData* icon = nil;
    if (sv_image && SvOK(sv_image)) {
        ptr = SvPV(sv_image, len);
        NSString* iconFile = [NSString stringWithUTF8String:ptr];
        NSURL* url = [NSURL URLWithString:iconFile];
        icon = [NSData dataWithContentsOfURL:url];
    }

    SV** svp;
    AV* av;
    int i, alen;

    NSMutableArray* all = [NSMutableArray array];
    av     = (AV*)SvRV(sv_allNotif);
    alen = av_len(av) + 1;
    for (i = 0; i < alen; ++i) {
        svp = av_fetch(av, i, 0);
        if (svp) {
            ptr = SvPV(*svp, len);
            [all addObject:[NSString stringWithUTF8String:ptr]];
        }
    }

    NSMutableArray* defaults = [NSMutableArray array];
    av     = (AV*)SvRV(sv_defaultNotif);
    alen = av_len(av) + 1;
    for (i = 0; i < alen; ++i) {
        svp = av_fetch(av, i, 0);
        if (svp) {
            ptr = SvPV(*svp, len);
            [defaults addObject:[NSString stringWithUTF8String:ptr]];
        }
    }

    NSDictionary* info;
    if (icon) {
        info = [NSDictionary dictionaryWithObjectsAndKeys:
                                            [NSNumber numberWithInt:1],     GROWL_TICKET_VERSION,
                                            @"org.unknownplace.cocoagrowl", GROWL_APP_ID,
                                            appName,                        GROWL_APP_NAME,
                                            all,                            GROWL_NOTIFICATIONS_ALL,
                                            defaults,                       GROWL_NOTIFICATIONS_DEFAULT,
                                            icon,                           GROWL_APP_ICON,
                                       nil];
    }
    else {
        info = [NSDictionary dictionaryWithObjectsAndKeys:
                                            [NSNumber numberWithInt:1],     GROWL_TICKET_VERSION,
                                            @"org.unknownplace.cocoagrowl", GROWL_APP_ID,
                                            appName,                        GROWL_APP_NAME,
                                            all,                            GROWL_NOTIFICATIONS_ALL,
                                            defaults,                       GROWL_NOTIFICATIONS_DEFAULT,
                                       nil];
    }

    [[Growl sharedInstance] setInfo:info];

    [pool drain];

    XSRETURN(0);
}

XS(growl_notify) {
    dXSARGS;

    if (items < 3) {
        Perl_croak(aTHX_ "Usage: _growl_notify($title, $description, $name[, $icon, $click_callback, $timeout_callback, $sticky, $priority])");
    }

    SV* sv_title = ST(0);
    SV* sv_desc  = ST(1);
    SV* sv_name  = ST(2);

    SV* sv_icon = NULL;
    if (items >= 4 && SvOK(ST(3))) sv_icon = ST(3);

    SV* sv_click_callback = NULL;
    SV* sv_timeout_callback = NULL;
    if (items >= 5 && SvOK(ST(4))) sv_click_callback = ST(4);
    if (items >= 6 && SvOK(ST(5))) sv_timeout_callback = ST(5);

    SV* sv_sticky = NULL;
    if (items >= 7 && SvOK(ST(6))) sv_sticky = ST(6);

    SV* sv_priority = NULL;
    if (items >= 8 && SvOK(ST(7))) sv_priority = ST(7);

    STRLEN len;
    char* s;

    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];

    s = SvPV(sv_title, len);
    NSString* title = [NSString stringWithUTF8String:s];

    s = SvPV(sv_desc, len);
    NSString* description = [NSString stringWithUTF8String:s];

    s = SvPV(sv_name, len);
    NSString* name = [NSString stringWithUTF8String:s];

    [GrowlAppBridge setGrowlDelegate:[Growl sharedInstance]];

    NSString* context_key = nil;
    if (sv_click_callback || sv_timeout_callback) {
        GrowlContext* context = [[GrowlContext alloc] init];
        context->click_cb = sv_click_callback;
        context->timeout_cb = sv_timeout_callback;

        if (sv_click_callback) SvREFCNT_inc(sv_click_callback);
        if (sv_timeout_callback) SvREFCNT_inc(sv_timeout_callback);

        context_key = [[Growl sharedInstance] addContext:context];
    }

    NSData* icon = nil;
    if (sv_icon && SvOK(sv_icon)) {
        s = SvPV(sv_icon, len);
        NSString* iconFile = [NSString stringWithUTF8String:s];
        NSURL* url = [NSURL URLWithString:iconFile];
        icon = [NSData dataWithContentsOfURL:url];
    }

    BOOL sticky = NO;
    if (sv_sticky && SvIV(sv_sticky)) sticky = YES;

    int priority = 0;
    if (sv_priority) priority = SvIV(sv_priority);

    [GrowlAppBridge notifyWithTitle:title
                        description:description
                   notificationName:name
                           iconData:icon
                           priority:priority
                           isSticky:sticky
                       clickContext:context_key];

    [pool drain];

    XSRETURN(0);
}

XS(boot_Cocoa__Growl) {
    SV* dir = get_sv("Cocoa::Growl::FRAMEWORK_DIR", FALSE);

    STRLEN len;
    char* d = SvPV(dir, len);

    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];

    GrowlAppBridge = objc_getClass("GrowlApplicationBridge");
    if (!GrowlAppBridge) {
        NSBundle* growl = [NSBundle bundleWithPath:[NSString stringWithUTF8String:d]];
        NSError* err = nil;
        if (growl && [growl loadAndReturnError:&err]) {
            //NSLog(@"growl load success");
            GrowlAppBridge = objc_getClass("GrowlApplicationBridge");
        }
        else {
            if (nil != err) {
                Perl_croak(aTHX_ "Coudn't load Growl framework: %s\n",
                    [[err localizedDescription] UTF8String]);
            }
            else {
                Perl_croak(aTHX_ "Coudn't initialize Growl bundle\n");
            }
        }
    }

    [pool drain];

    newXS("Cocoa::Growl::growl_installed", growl_installed, __FILE__);
    newXS("Cocoa::Growl::growl_running", growl_running, __FILE__);
    newXS("Cocoa::Growl::_growl_register", growl_register, __FILE__);
    newXS("Cocoa::Growl::_growl_notify", growl_notify, __FILE__);
}


Makefile.PL
use inc::Module::Install;
use File::Spec;

{
    # check OS and version
    die 'OS unsupported' unless ($^O eq 'darwin');

    my @ver = map { chomp; $_ } split /\./, `sw_vers -productVersion`;
    my $ver = sprintf '%d.%03d%03d', @ver;

    die 'OS unsupported' unless $ver ge '10.005';
}

name 'Cocoa-Growl';
all_from 'lib/Cocoa/Growl.pm';
readme_from 'lib/Cocoa/Growl.pm';

requires 'parent';
requires 'URI';
requires 'File::ShareDir';

test_requires 'Test::More' => '0.88';
author_tests('xt');

auto_set_repository;

use_ppport '3.19';
cc_append_to_libs '-lobjc';

# i don't know why but without this occuered compilation error on OSX system perl
cc_append_to_ccflags '-D__QUICKDRAWAPI__';

my $use_local_framework;
if ($use_local_framework = (grep(/USE_LOCAL_GROWL_FRAMEWORK=.*/, @ARGV))[0] ) {
    $use_local_framework =~ s/.*?=//;
}
@ARGV = grep !/USE_LOCAL_GROWL_FRAMEWORK/, @ARGV;

if ($use_local_framework) {
    cc_append_to_ccflags '-DUSE_LOCAL_GROWL';

    makemaker_args->{dynamic_lib} = {
        OTHERLDFLAGS => '-framework Foundation -framework Growl',
    };
}
else {
    cc_append_to_ccflags sprintf '-I%s',
        File::Spec->rel2abs(File::Spec->catfile('share', 'Growl.framework', 'Versions', 'A', 'Headers'));

    makemaker_args->{dynamic_lib} = {
        OTHERLDFLAGS => '-framework Foundation',
    };

    install_share;
}

{
    # http://www.mail-archive.com/makemaker@perl.org/msg02823.html
    # Add some required machinery to support .m files
    package MY;
    sub c_o {
        my $inherited = shift->SUPER::c_o(@_);
        $inherited .= <<'EOMESS';
.m$(OBJ_EXT):
	$(CCCMD) $(CCCDLFLAGS) "-I$(PERL_INC)" $(PASTHRU_DEFINE) $(DEFINE) $*.m

EOMESS
        $inherited;
    }

    sub xs_c {
        my $inherited = shift->SUPER::xs_c(@_);
        $inherited .= <<'EOMESS';
.xs.m:
	$(XSUBPPRUN) $(XSPROTOARG) $(XSUBPPARGS) $(XSUBPP_EXTRA_ARGS) $*.xs > $*.xsm && $(MV) $*.xsm $*.m

EOMESS
        $inherited;
    }

    sub init_dirscan {
        my $self = shift;
        $self->SUPER::init_dirscan;
    
        foreach my $name ($self->lsdir($self->curdir)) {
            next unless $name =~ /\.m$/;
            push @{$self->{'C'}}, $name;
            $name =~ s/\.m$/$self->{'OBJ_EXT'}/;
            push @{$self->{'O_FILES'}}, $name;
        }
    }
}

WriteAll;


lib/Cocoa/Growl.pm
package Cocoa::Growl;
use strict;
use warnings;
use parent 'Exporter';

use XSLoader;
use URI;

our $VERSION = '0.04';

our @EXPORT_OK   = qw(growl_installed growl_running growl_register growl_notify);
our %EXPORT_TAGS = (all => \@EXPORT_OK);

use File::ShareDir ();
use File::Spec;

our $FRAMEWORK_DIR = do {
    (my $dist = __PACKAGE__) =~ s/::/-/g;
    File::Spec->catfile(File::ShareDir::dist_dir($dist), 'Growl.framework');
};

XSLoader::load __PACKAGE__, $VERSION;

sub growl_notify {
    my %info = @_;

    my $title       = $info{title} || '';
    my $description = $info{description} || '';
    my $notifName   = $info{name} || '';
    my $icon        = $info{icon};
    my $sticky      = $info{sticky};
    my $priority    = $info{priority} || 0;

    my $on_click   = $info{on_click};
    my $on_timeout = $info{on_timeout};

    if ($icon) {
        my $uri = URI->new($icon);
        $uri->scheme or do { $uri->scheme('file'); $uri->host('') };
        $icon = $uri->as_string;
    }

    _growl_notify($title, $description, $notifName, $icon, $on_click, $on_timeout, $sticky, $priority);
}

sub growl_register {
    my %info = @_;

    my $appName  = $info{app} || __PACKAGE__;
    my $appIcon  = $info{icon};
    my $all      = $info{notifications};
    my $defaults = $info{defaults} || $all;

    if ($appIcon) {
        my $uri = URI->new($appIcon);
        $uri->scheme or do { $uri->scheme('file'); $uri->host('') };
        $appIcon = $uri->as_string;
    }

    _growl_register($appName, $all, $defaults, $appIcon);
}

1;

__END__

=for stopwords NSRunLoop Str AnyEvent

=head1 NAME

Cocoa::Growl - Yet another growl module using Growl.framework

=head1 SYNOPSIS

    use Cocoa::Growl ':all';
    
    my $installed = growl_installed(); # true if Growl is installed.
    my $running   = growl_running();   # true if Growl is running.
    
    # register application
    growl_register(
        app           => 'My growl script',
        icon          => '/path/to/icon.png', # or 'http://url/to/icon'
        notifications => [qw(Notification1 Notification2)],
    );
    
    # show growl notification
    growl_notify(
        name        => 'Notification1',
        title       => 'Hello!',
        description => 'Growl world!',
    );

=head1 DESCRIPTION

=head1 FUNCTIONS

No function is exported by default, but all functions is exportable.
And ':all' tag export all functions.

=head2 growl_installed

    my $installed = growl_installed();

Return true value if growl is installed.

=head2 growl_running

    my $running = growl_running();

Return true value if growl is running.

=head2 growl_register(%parameters)

Register application to growl.

    growl_register(
        app           => 'My growl script',
        icon          => '/path/to/icon.png', # or 'http://url/to/icon'
        notifications => [qw(Notification1 Notification2)],
    );

Available parameters are:

=over 4

=item * app => 'Str' (Required)

The name of the application. 
This is listed in Growl preference panel.

=item * icon => 'Str'

Application icon image path or URL. This image is showed in Growl preference panel, and used notification default image.

=item * notifications => 'ArrayRef' (Required)

List of notification names.
These names will be displayed in Growl preference pane to let users customize options for each notification.

=item * defaults => 'ArrayRef'

List of notification names to enable by default.
If this parameter is not set, all notifications is to become default.

=back

=head2 growl_notify(%parameters)

Show growl notify.

    growl_notify(
        name        => 'Notification1',
        title       => 'Hello!',
        description => 'Growl world!',
    );

Available options are:

=over 4

=item * name => 'Str' (Required)

The internal name of the notification. Should be human-readable, as it will be displayed in the Growl preference pane.
And this value is required to be registered by C<growl_register> before calling this function.

=item * title => 'Str'

The title of the notification displayed to the user.

=item * description => 'Str'

The full description of the notification displayed to the user.

=item * icon => 'Str'

Image file path or URL to show with the notification as its icon. If this value is not set, the application's icon will be used instead.

=item * sticky => 'Bool'

If true value is set, the notification will remain on screen until clicked.
Not all Growl displays support sticky notifications.

=item * priority => 'Int'

The priority of the notification. The default value is 0; positive values are higher priority and negative values are lower priority.
Not all Growl displays support priority.

=item * on_click => 'CodeRef',

This callback is called when notification is clicked.
See also CALLBACK NOTICE below.

=item * on_timeout => 'CodeRef',

This callback is called when notification is timeout. (also called notification closed by close button)

=back

=head3 CALLBACK NOTICE

You should run Cocoa's event loop NSRunLoop to be enable callbacks.
Simplest way to do that is use this module with L<Cocoa::EventLoop>.

    use Cocoa::EventLoop;
    use Cocoa::Growl ':all';
    
    growl_register(
        name          => 'test script',
        notifications => ['test notification'],
    );
    
    my $wait = 1;
    growl_notify(
        name        => 'test notification',
        title       => 'Hello',
        description => 'Growl World!',
        onClick => sub {
            warn 'click';
            $wait = 0;
        },
        onTimeout => sub {
            warn 'timeout';
            $want = 0;
        },
    );
    
    Cocoa::EventLoop->run_while(0.1) while unless $wait;

If you want to write more complicated script, use L<AnyEvent> and L<AnyEvent::Impl::NSRunLoop>.
L<AnyEvent::Impl::NSRunLoop> is a wrapper for L<AnyEvent> and L<Cocoa::EventLoop> and by using this module, you can use cocoa's event loop transparently in your AnyEvent application.

This is little example:

    use AnyEvent;
    use AnyEvent::Impl::NSRunLoop;
    
    use Cocoa::Growl ':all';;
    
    my $cv = AnyEvent->condvar;
    
    growl_register(
        name          => 'test script',
        notifications => ['test notification'],
    );
    
    growl_notify(
        name        => 'test notification',
        title       => 'Hello',
        description => 'Growl World!',
        onClick => sub {
            warn 'click';
            $cv->send;
        },
        onTimeout => sub {
            warn 'timeout';
            $cv->send;
        },
    );
    
    $cv->recv;

This script show one notification and wait until notification closed.

=head1 USE YOUR OWN Growl.framework

Although this module bundle Growl.framework and load it by default, you can load your own Growl.framework.
To do that, save your Growl.framework to C</Library/Frameworks/Growl.framework/>, and add C<USE_LOCAL_GROWL_FRAMEWORK=1> option when run Makefile.PL

    perl Makefile.PL USE_LOCAL_GROWL_FRAMEWORK=1


=head1 AUTHOR

Daisuke Murase <typester@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2010 by KAYAC Inc.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut
